File: python/maze/generator/AldousBroder.py
import random
from maze.Grid import Grid


def generate(height: int, width: int) -> Grid:
    """
    Aldous-Broder maze generation. Aldous-Broder generates perfect mazes, (mazes without loops), using "random-walks".
    This avoids creating mazes with biases (like Binary Tree) and instead produces mazes with lots of winding passages.

    The idea behind it is as follows:
    1. Start anywhere in the grid you want, and choose a random neighbor.
    2. Move to that neighbor, and if it has not previously been visited, link it to the prior cell.
    3. Repeat until every cell has been visited.
    :param height:
    :param width:
    :return:
    """

    grid = Grid(height, width)
    # the current cell being visited
    current = grid.random_cell()
    unvisited_count = grid.size() - 1

    while unvisited_count > 0:
        # choose a random neighbor of the current cell
        rand_neighbor = random.choice(current.neighbors())

        # if the random neighbor is not linked to anything, then link current to it
        if not rand_neighbor.linked_cells():
            current.link(rand_neighbor)
            unvisited_count -= 1

        # make rand_neighbor the new current cell
        current = rand_neighbor

    return grid

File: java/src/main/java/com/mazes/generator/AldousBroder.java
package com.mazes.generator;

import com.mazes.Cell;
import com.mazes.Grid;

import java.util.List;
import java.util.Random;

/**
 * Aldous-Broder maze generation. Aldous-Broder generates perfect mazes, (mazes without loops), using "random-walks".
 * This avoids creating mazes with biases (like Binary Tree) and instead produces mazes with lots of winding passages.
 *
 * The idea behind it is as follows:
 * 1. Start anywhere in the grid you want, and choose a random neighbor.
 * 2. Move to that neighbor, and if it has not previously been visited, link it to the prior cell.
 * 3. Repeat until every cell has been visited.
 */
public class AldousBroder {

    Random random = new Random();

    public Grid generate(int height, int width) {
        Grid grid = new Grid(height, width);

        Cell current = grid.randomCell();
        int unvisitedCount = grid.size() - 1;

        while (unvisitedCount > 0) {
            // pick a random neighbor of the current cell
            List<Cell> neighbors = current.neighbors();
            Cell neighbor = neighbors.get(random.nextInt(neighbors.size()));

            if (neighbor.links().isEmpty()) {
                // link to the neighbor
                current.link(neighbor, true);
                unvisitedCount--;
            }
            // neighbor becomes the current cell
            current = neighbor;
        }

        return grid;
    }
}
